crates/rpc/src/methods/token_add.rs

```diff
@@ -0,0 +1,317 @@
+            }
+        }
+        Ok(())
+    }
+
+    pub fn check_decimals(&self) -> Result<()> {
+        if let Some(erc20_token) = &self.erc20_token {
+            let decimals = erc20_token.decimals.unwrap_or(0);
+            if decimals > 36 {
+                return Err(Error::DecimalsInvalid(decimals));
+            }
+        }
+        Ok(())
+    }
+
+    pub async fn check_network(&self) -> Result<()> {
+        let current_chain_id = self.get_current_chain_id().await?;
+        let chain_id = self.chain_id.unwrap_or(current_chain_id);
+        if current_chain_id != chain_id {
+            return Err(Error::NetworkInvalid);
+        }
+        Ok(())
+    }
+
+    pub async fn run(self) -> Result<()> {
+        self.check_type()?;
+        self.check_network().await?;
+
+        match self._type.as_str() {
+            "ERC20" => {
+                self.check_symbol()?;
+                self.check_decimals()?;
+                // TODO: check if 'type' from call matches the added 'tokenType'
+            }
+            "ERC721" => {
+                // TODO: check if 'type' from call matches the added 'tokenType'
+                // TODO: check if wallet is the owner of the token
+            }
+            "ERC1155" => {
+                // TODO: check if 'type' from call matches the added 'tokenType'
+                // TODO: check if wallet is the owner of the token
+            }
+            _ => return Err(Error::TypeInvalid(self._type.clone())),
+        }
+
+        let dialog = match self._type.as_str() {
+            "ERC20" => Dialog::new(
+                "erc20token-add",
+                serde_json::to_value(&self.erc20_token).unwrap(),
+            ),
+            "ERC721" => Dialog::new(
+                "erc721token-add",
+                serde_json::to_value(&self.erc721_token).unwrap(),
+            ),
+            "ERC1155" => Dialog::new(
+                "erc1155token-add",
+                serde_json::to_value(&self.erc1155_token).unwrap(),
+            ),
+            _ => return Err(Error::TypeInvalid(self._type.clone())),
+        };
+        dialog.open().await?;
+
+        while let Some(msg) = dialog.recv().await {
+            match msg {
+                DialogMsg::Data(msg) => {
+                    if let Some("accept") = msg.as_str() {
+                        self.on_accept().await?;
+                        break;
+                    }
+                }
+
+                DialogMsg::Close => return Err(Error::UserRejected),
+            }
+        }
+
+        Ok(())
+    }
+
+    pub async fn on_accept(&self) -> Result<()> {
+        let current_wallet_address = self.get_current_wallet_address().await?;
+        let current_chain_id = self.get_current_chain_id().await?;
+        match self._type.as_str() {
+            "ERC20" => {
+                if let Some(erc20_token) = &self.erc20_token {
+                    print!("\nAdd ERC20: {}\n", erc20_token.address);
```

Comment:
not sure what the purpose of these prints are. if you want to log to the console, check the `tracing::info!` macro (or `tracing::debug`, etc)

Reply:
This is temporary just to log a final confirmation, it will be removed

Reply:
temporary until when?
Don't get me wrong, I'm fine with reviewing in-progress code, but the PR should indicate so (usually by setting it as a draft).

And since you explicitly asked for review, it should be clear what is finished and what's not

Reply:
Understood! I had all the implementation for the EIP ready but felt like it could be broken into different PR's/steps so it would be easier to review. This is why there are some unfinished parts, but as your said they should have been clearly explained.